### DISCUSSION TITLE
Remove node with filter

### DISCUSSION THREAD

**Discussion by nathan30** [downforcetech/fluidxml.php#18] [Questions]
Hi,

I want to remove some nodes with a certain end and startdate. Do you have any clue to achieve this ?

Thanks in advance

**Comment by daniele-orlando** [comment]
Hi nathan,

```
$doc('//any/query')
    ->filter(function ($i, $node) {
        return <true_or_false>;                                                                            
    })->remove();  
```

**Comment by nathan30** [comment]
Thanks for your answer and sorry for the late answer. With your piece of code I didn't achieve what I want to do..

I have an XML file organize like this : 

```
<?xml version="1.0" encoding="UTF-8"?>
<monthly>
  <event>
    <name>Indisponible</name>
    <startdate>2016-07-25</startdate>
    <enddate>2016-08-05</enddate>
    <starttime/>
    <endtime/>
    <color>#ffb128</color>
    <url/>
  </event>
  <event>
    <name>Indisponible</name>
    <startdate>2016-08-10</startdate>
    <enddate>2016-08-28</enddate>
    <starttime/>
    <endtime/>
    <color>#ffb128</color>
    <url/>
  </event>
  <event>
    <name>Indisponible</name>
    <startdate>2016-09-06</startdate>
    <enddate>2016-09-08</enddate>
    <starttime/>
    <endtime/>
    <color>#ffb128</color>
    <url/>
  </event>
  <event>
    <name>Indisponible</name>
    <startdate>2016-09-06</startdate>
    <enddate>2016-09-07</enddate>
    <starttime/>
    <endtime/>
    <color>#ffb128</color>
    <url/>
  </event>
</monthly>
```

And I want to know how to delete one of the "event" node. I try something with your answer, but none of my I try works...

Thanks in advance

**Comment by daniele-orlando** [comment]
Supposing you want to delete the `event` node with `startdate == '2016-07-25'`, you don't need to filter:

```
// An 'event' node with a child 'startdate' having '2016-07-25' as content.
$doc('//event[startdate[text()="2016-07-25"]]')->remove();
```

or

```
$doc('//startdate[text()="2016-07-25"]]')  // We match the 'startdate' node.
    ->query('..') // We select the parent, the 'event' node to remove.
    ->remove();
```

If you really want to use filter:

```
$doc('//startdate')
    ->filter(function ($i, $node) {
        return $node->nodeValue === '2016-07-25';
    })
    ->query('..')
    ->remove();
```

Hope you found what you were looking for.
Dan

**Comment by nathan30** [comment]
Awesome, thanks a lot for your amazing help

EDIT :  I try the first one, and the other, and it's not working.. I have no error but nothing was deleted.. Here is my piece of code : 

`require_once 'FluidXml.php';

use \FluidXml\FluidXml;
use \FluidXml\FluidNamespace;
use function \FluidXml\fluidxml;
use function \FluidXml\fluidns;
use function \FluidXml\fluidify;

$split = array();
$doc1 = fluidify('events.xml',['root' => 'montlhy']);
// Listing the events in an array
$doc1->query('//event')
        ->each(function($i, $domnode) {
                $this->attr('id', $i + 1);
                $split[$i] = explode("Indisponible",$this);
                print_r($split);
                //echo $domnode->nodeValue;
     });
$doc1("//event[startdate[text()='2016-07-25']]")->remove();`

**Comment by daniele-orlando** [comment]
I tested your code and FluidXML behaves as expected removing the first `event` node, the one with `startdate = 2016-07-25`.

`echo $doc1;`

outputs this

```
<?xml version="1.0" encoding="UTF-8"?>
<monthly>
  <event id="2">
    <name>Indisponible</name>
    <startdate>2016-08-10</startdate>
    <enddate>2016-08-28</enddate>
    <starttime/>
    <endtime/>
    <color>#ffb128</color>
    <url/>
  </event>
  <event id="3">
    <name>Indisponible</name>
    <startdate>2016-09-06</startdate>
    <enddate>2016-09-08</enddate>
    <starttime/>
    <endtime/>
    <color>#ffb128</color>
    <url/>
  </event>
  <event id="4">
    <name>Indisponible</name>
    <startdate>2016-09-06</startdate>
    <enddate>2016-09-07</enddate>
    <starttime/>
    <endtime/>
    <color>#ffb128</color>
    <url/>
  </event>
</monthly>
```

p.s.: `fluidify` function takes only one argument, the input file. `['root' => 'montlhy']` is ignored.

**Comment by nathan30** [comment]
Well, I really don't understand why I don't have the same behavior as you.. I try different things and I always have the same result with no node deleted

**Comment by daniele-orlando** [comment]
Attach your files here, XML and PHP code. I'll give it a look.

**Comment by nathan30** [comment]
Thanks a lot 

[fluidXML.zip](https://github.com/servo-php/fluidxml/files/516944/fluidXML.zip)

**Comment by daniele-orlando** [comment]
Sorry for the late reply. I checked up your code and the call at line `135`

```
$doc1("//event[color[text()='#ffb128']]")->remove();
```

is behaving as expected, removing every event node with the searched color. You can see it simply adding this code after line `135`:

```
echo $doc1;
```

which will print an empty `<monthly/>` tag.

I think you have an error in the logic of your code, maybe  missing the `->save('events.xml')` call after removing the events.

Cheers!

**Comment by nathan30** [comment]
Awesome,

Thanks a lot for your help, really

Enjoy your day
